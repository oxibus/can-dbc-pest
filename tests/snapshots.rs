#![cfg(feature = "encoding")]

use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [("tests/shared-test-files/dbc-cantools", "dbc-cantools")];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match DbcParser::parse(Rule::file, &buffer) {
            Ok(pairs) => assert_debug_snapshot!(file_name, pairs),
            Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
        }
    } else {
        panic!("Failed to decode {file_name}.dbc as cp1252");
    }
}
