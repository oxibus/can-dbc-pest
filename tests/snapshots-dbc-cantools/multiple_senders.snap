---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"\"\n\n\nNS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_: FOO BAR FIE FUM\n\n\nBO_ 1 Bar: 4 FOO\n SG_ Binary32 : 0|32@1- (1,0) [0|0] \"\"  FUM\n\nBO_TX_BU_ 1 : FOO,BAR,FIE;\n\n\n\nBA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;\nBA_DEF_DEF_  \"GenMsgCycleTime\" 0;\n\n\n\n\n\n",
            start: 0,
            end: 541,
        },
        inner: [
            Pair {
                rule: line,
                span: Span {
                    str: "VERSION \"\"",
                    start: 0,
                    end: 10,
                },
                inner: [
                    Pair {
                        rule: version,
                        span: Span {
                            str: "VERSION \"\"",
                            start: 0,
                            end: 10,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"\"",
                                    start: 8,
                                    end: 10,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 10,
                    end: 11,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 11,
                    end: 12,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 12,
                    end: 13,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "NS_ : ",
                    start: 13,
                    end: 19,
                },
                inner: [
                    Pair {
                        rule: new_symbols,
                        span: Span {
                            str: "NS_ : ",
                            start: 13,
                            end: 19,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 19,
                    end: 20,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "NS_DESC_\n",
                    start: 21,
                    end: 30,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "NS_DESC_",
                            start: 21,
                            end: 29,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CM_\n",
                    start: 31,
                    end: 35,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "CM_",
                            start: 31,
                            end: 34,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_\n",
                    start: 36,
                    end: 44,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_",
                            start: 36,
                            end: 43,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_\n",
                    start: 45,
                    end: 49,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_",
                            start: 45,
                            end: 48,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "VAL_\n",
                    start: 50,
                    end: 55,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "VAL_",
                            start: 50,
                            end: 54,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CAT_DEF_\n",
                    start: 56,
                    end: 65,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 56,
                            end: 64,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CAT_\n",
                    start: 66,
                    end: 71,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "CAT_",
                            start: 66,
                            end: 70,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "FILTER\n",
                    start: 72,
                    end: 79,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "FILTER",
                            start: 72,
                            end: 78,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_DEF_\n",
                    start: 80,
                    end: 92,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 80,
                            end: 91,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "EV_DATA_\n",
                    start: 93,
                    end: 102,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "EV_DATA_",
                            start: 93,
                            end: 101,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "ENVVAR_DATA_\n",
                    start: 103,
                    end: 116,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 103,
                            end: 115,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SGTYPE_\n",
                    start: 117,
                    end: 125,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SGTYPE_",
                            start: 117,
                            end: 124,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SGTYPE_VAL_\n",
                    start: 126,
                    end: 138,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 126,
                            end: 137,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_SGTYPE_\n",
                    start: 139,
                    end: 154,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 139,
                            end: 153,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_SGTYPE_\n",
                    start: 155,
                    end: 166,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 155,
                            end: 165,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIG_TYPE_REF_\n",
                    start: 167,
                    end: 181,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 167,
                            end: 180,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "VAL_TABLE_\n",
                    start: 182,
                    end: 193,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 182,
                            end: 192,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIG_GROUP_\n",
                    start: 194,
                    end: 205,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 194,
                            end: 204,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIG_VALTYPE_\n",
                    start: 206,
                    end: 219,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 206,
                            end: 218,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIGTYPE_VALTYPE_\n",
                    start: 220,
                    end: 237,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 220,
                            end: 236,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BO_TX_BU_\n",
                    start: 238,
                    end: 248,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 238,
                            end: 247,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_REL_\n",
                    start: 249,
                    end: 261,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 249,
                            end: 260,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_REL_\n",
                    start: 262,
                    end: 270,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_REL_",
                            start: 262,
                            end: 269,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_DEF_REL_\n",
                    start: 271,
                    end: 287,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 271,
                            end: 286,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_SG_REL_\n",
                    start: 288,
                    end: 299,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 288,
                            end: 298,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_EV_REL_\n",
                    start: 300,
                    end: 311,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 300,
                            end: 310,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_BO_REL_\n",
                    start: 312,
                    end: 323,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 312,
                            end: 322,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SG_MUL_VAL_\n",
                    start: 324,
                    end: 336,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 324,
                            end: 335,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 336,
                    end: 337,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BS_:",
                    start: 337,
                    end: 341,
                },
                inner: [
                    Pair {
                        rule: bit_timing,
                        span: Span {
                            str: "BS_:",
                            start: 337,
                            end: 341,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 341,
                    end: 342,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 342,
                    end: 343,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_: FOO BAR FIE FUM",
                    start: 343,
                    end: 363,
                },
                inner: [
                    Pair {
                        rule: nodes,
                        span: Span {
                            str: "BU_: FOO BAR FIE FUM",
                            start: 343,
                            end: 363,
                        },
                        inner: [
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "FOO",
                                    start: 348,
                                    end: 351,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "BAR",
                                    start: 352,
                                    end: 355,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "FIE",
                                    start: 356,
                                    end: 359,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "FUM",
                                    start: 360,
                                    end: 363,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 363,
                    end: 364,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 364,
                    end: 365,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 365,
                    end: 366,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BO_ 1 Bar: 4 FOO",
                    start: 366,
                    end: 382,
                },
                inner: [
                    Pair {
                        rule: message,
                        span: Span {
                            str: "BO_ 1 Bar: 4 FOO",
                            start: 366,
                            end: 382,
                        },
                        inner: [
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "1",
                                    start: 370,
                                    end: 371,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "Bar",
                                    start: 372,
                                    end: 375,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "4",
                                    start: 377,
                                    end: 378,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "FOO",
                                    start: 379,
                                    end: 382,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 382,
                    end: 383,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SG_ Binary32 : 0|32@1- (1,0) [0|0] \"\"  FUM",
                    start: 384,
                    end: 426,
                },
                inner: [
                    Pair {
                        rule: signal,
                        span: Span {
                            str: "SG_ Binary32 : 0|32@1- (1,0) [0|0] \"\"  FUM",
                            start: 384,
                            end: 426,
                        },
                        inner: [
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "Binary32",
                                    start: 388,
                                    end: 396,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "0",
                                    start: 399,
                                    end: 400,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "32",
                                    start: 401,
                                    end: 403,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "1",
                                    start: 404,
                                    end: 405,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "1",
                                    start: 408,
                                    end: 409,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "0",
                                    start: 410,
                                    end: 411,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "0",
                                    start: 414,
                                    end: 415,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "0",
                                    start: 416,
                                    end: 417,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"\"",
                                    start: 419,
                                    end: 421,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "FUM",
                                    start: 423,
                                    end: 426,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 426,
                    end: 427,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 427,
                    end: 428,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BO_TX_BU_ 1 : FOO,BAR,FIE;",
                    start: 428,
                    end: 454,
                },
                inner: [
                    Pair {
                        rule: bo_tx_bu,
                        span: Span {
                            str: "BO_TX_BU_ 1 : FOO,BAR,FIE;",
                            start: 428,
                            end: 454,
                        },
                        inner: [
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "1",
                                    start: 438,
                                    end: 439,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "FOO",
                                    start: 442,
                                    end: 445,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "BAR",
                                    start: 446,
                                    end: 449,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "FIE",
                                    start: 450,
                                    end: 453,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 454,
                    end: 455,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 455,
                    end: 456,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 456,
                    end: 457,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 457,
                    end: 458,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;",
                    start: 458,
                    end: 501,
                },
                inner: [
                    Pair {
                        rule: attr_def,
                        span: Span {
                            str: "BA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;",
                            start: 458,
                            end: 501,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"GenMsgCycleTime\"",
                                    start: 471,
                                    end: 488,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "0",
                                    start: 493,
                                    end: 494,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "65535",
                                    start: 495,
                                    end: 500,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 501,
                    end: 502,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_DEF_  \"GenMsgCycleTime\" 0;",
                    start: 502,
                    end: 535,
                },
                inner: [
                    Pair {
                        rule: ba_def_def,
                        span: Span {
                            str: "BA_DEF_DEF_  \"GenMsgCycleTime\" 0;",
                            start: 502,
                            end: 535,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"GenMsgCycleTime\"",
                                    start: 515,
                                    end: 532,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "0",
                                    start: 533,
                                    end: 534,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 535,
                    end: 536,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 536,
                    end: 537,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 537,
                    end: 538,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 538,
                    end: 539,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 539,
                    end: 540,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 540,
                    end: 541,
                },
                inner: [],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 541,
                    end: 541,
                },
                inner: [],
            },
        ],
    },
]
